homeassistant/components/abode.py

```diff
@@ -65,47 +133,149 @@ def setup(hass, config):
+        if not hass.data[DOMAIN].polling:
+            hass.data[DOMAIN].abode.events.start()
+
     def logout(event):
         """Logout of Abode."""
-        abode.stop_listener()
-        abode.logout()
+        if not hass.data[DOMAIN].polling:
+            hass.data[DOMAIN].abode.events.stop()
+
+        hass.data[DOMAIN].abode.logout()
         _LOGGER.info("Logged out of Abode")
 
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, startup)
     hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, logout)
 
-    def startup(event):
-        """Listen for push events."""
-        abode.start_listener()
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_START, startup)
+def setup_abode_events(hass):
+    """Event callbacks."""
+    import abodepy.helpers.timeline as TIMELINE
 
-    return True
+    def event_callback(event_json, event=None):
+        """Handle an event callback from Abode."""
+        if event:
+            data = {
+                ATTR_DEVICE_ID: event_json.get(ATTR_DEVICE_ID, ''),
+                ATTR_DEVICE_NAME: event_json.get(ATTR_DEVICE_NAME, ''),
+                ATTR_DEVICE_TYPE: event_json.get(ATTR_DEVICE_TYPE, ''),
+                ATTR_EVENT_CODE: event_json.get(ATTR_EVENT_CODE, ''),
+                ATTR_EVENT_NAME: event_json.get(ATTR_EVENT_NAME, ''),
+                ATTR_EVENT_TYPE: event_json.get(ATTR_EVENT_TYPE, ''),
+                ATTR_EVENT_UTC: event_json.get(ATTR_EVENT_UTC, ''),
+                ATTR_USER_NAME: event_json.get(ATTR_USER_NAME, ''),
+                ATTR_DATE: event_json.get(ATTR_DATE, ''),
+                ATTR_TIME: event_json.get(ATTR_TIME, ''),
+            }
+            _LOGGER.info(data)
+            hass.bus.fire(event, data)
+
+    events = [TIMELINE.ALARM_GROUP, TIMELINE.ALARM_END_GROUP,
+              TIMELINE.PANEL_FAULT_GROUP, TIMELINE.PANEL_RESTORE_GROUP,
+              TIMELINE.AUTOMATION_GROUP]
+
+    for event in events:
+        hass.data[DOMAIN].abode.events.add_event_callback(
+            event,
+            partial(event_callback, event=event))
 
 
 class AbodeDevice(Entity):
     """Representation of an Abode device."""
 
-    def __init__(self, controller, device):
+    def __init__(self, data, device):
         """Initialize a sensor for Abode device."""
-        self._controller = controller
+        self._data = data
         self._device = device
 
     @asyncio.coroutine
     def async_added_to_hass(self):
         """Subscribe Abode events."""
         self.hass.async_add_job(
-            self._controller.register, self._device,
-            self._update_callback
+            self._data.abode.events.add_device_callback,
+            self._device.device_id, self._update_callback
         )
 
     @property
     def should_poll(self):
         """Return the polling state."""
+        if self._data.polling:
+            return True
+
         return False
 
+    def update(self):
+        """Update automation state."""
+        self._device.refresh()
```

Comment:
Is `refresh()` throttled or will it be called once per update cycle per device?

Reply:
It is not throttled. I just added back how the polling update worked in the original PR #9030. Should it be done differently?